baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/PreferCollectionConstructors.java

```diff
@@ -0,0 +1,345 @@
+            ImmutableMap.<Matcher<ExpressionTree>, Class<?>>builder()
+                    .put(NEW_ARRAY_LIST, ArrayList.class)
+                    .put(NEW_ARRAY_LIST_WITH_ITERABLE, ArrayList.class)
+                    .put(NEW_ARRAY_LIST_WITH_CAPACITY, ArrayList.class)
+                    .put(NEW_LINKED_LIST, LinkedList.class)
+                    .put(NEW_LINKED_LIST_WITH_ITERABLE, LinkedList.class)
+                    .put(NEW_COPY_ON_WRITE_ARRAY_LIST, CopyOnWriteArrayList.class)
+                    .put(NEW_COPY_ON_WRITE_ARRAY_LIST_WITH_ITERABLE, CopyOnWriteArrayList.class)
+                    .put(NEW_CONCURRENT_MAP, ConcurrentHashMap.class)
+                    .put(NEW_HASH_MAP, HashMap.class)
+                    .put(NEW_HASH_MAP_WITH_MAP, HashMap.class)
+                    .put(NEW_HASH_MAP_WITH_EXPECTED_SIZE, HashMap.class)
+                    .put(NEW_COPY_ON_WRITE_ARRAY_SET, CopyOnWriteArraySet.class)
+                    .put(NEW_COPY_ON_WRITE_ARRAY_SET_WITH_ITERABLE, CopyOnWriteArraySet.class)
+                    .put(NEW_LINKED_HASH_SET, LinkedHashSet.class)
+                    .put(NEW_LINKED_HASH_SET_WITH_ITERABLE, LinkedHashSet.class)
+                    .put(NEW_LINKED_HASH_SET_WITH_EXPECTED_SIZE, LinkedHashSet.class)
+                    .put(NEW_TREE_SET, TreeSet.class)
+                    .put(NEW_TREE_SET_WITH_COMPARATOR, TreeSet.class)
+                    .put(NEW_TREE_SET_WITH_ITERABLE, TreeSet.class)
+                    .put(NEW_HASH_SET, HashSet.class)
+                    .put(NEW_HASH_SET_WITH_ITERABLE, HashSet.class)
+                    .put(NEW_HASH_SET_WITH_EXPECTED_SIZE, HashSet.class)
+                    .put(NEW_TREE_MAP, TreeMap.class)
+                    .put(NEW_TREE_MAP_WITH_SORTED_MAP, TreeMap.class)
+                    .put(NEW_TREE_MAP_WITH_COMPARATOR, TreeMap.class)
+                    .put(NEW_LINKED_HASH_MAP, LinkedHashMap.class)
+                    .put(NEW_LINKED_HASH_MAP_WITH_MAP, LinkedHashMap.class)
+                    .put(NEW_LINKED_HASH_MAP_WITH_EXPECTED_SIZE, LinkedHashMap.class)
+                    .put(NEW_ENUM_MAP, EnumMap.class)
+                    .put(NEW_ENUM_MAP_WITH_CLASS, EnumMap.class)
+                    .put(NEW_ENUM_MAP_WITH_MAP, EnumMap.class)
+                    .put(NEW_IDENTITY_HASH_MAP, IdentityHashMap.class)
+                    .build();
+
+    private static final Set<Matcher<ExpressionTree>> requiresCollectionArg = ImmutableSet.of(
+            NEW_ARRAY_LIST_WITH_ITERABLE,
+            NEW_LINKED_LIST_WITH_ITERABLE,
+            NEW_COPY_ON_WRITE_ARRAY_LIST_WITH_ITERABLE,
+            NEW_COPY_ON_WRITE_ARRAY_SET_WITH_ITERABLE,
+            NEW_LINKED_HASH_SET_WITH_ITERABLE,
+            NEW_TREE_SET_WITH_ITERABLE,
+            NEW_HASH_SET_WITH_ITERABLE);
+
+    @Override
+    public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+
+        Class<?> collectionClass = findCollectionClassToUse(state, tree);
```

Comment:
Might want to create a static fast check using Matchers.anyOf(classMap.keySet()) to avoid creating an iterable for every method invocation.

Reply:
The `ImmutableMap` seems to internally only create a key set once anyway, right? And the `anyOff(...)` matcher still just loops over the matchers. Maybe I'm missing something, but it seems like it wouldn't make a difference?

Reply:
Ah right, I thought matchers.anyOf did something cleverer :-/